src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs

```diff
@@ -313,68 +356,121 @@ private ObjectPropertyInitializer[] GetPropertyInitializers(IReadOnlyList<Parame
 
         protected override CSharpType? CreateInheritedType()
         {
-            if (_inputModel.BaseModel is not null)
-                return _typeFactory.CreateType(_inputModel.BaseModel!);
+            if (GetSourceBaseType() is {} sourceBaseType && _typeFactory.TryCreateType(sourceBaseType, out CSharpType? baseType))
+            {
+                return baseType;
+            }
 
-            return null;
+            return _inputModel.BaseModel is {} baseModel
+                ? _typeFactory.CreateType(baseModel)
+                : null;
         }
 
+        private INamedTypeSymbol? GetSourceBaseType()
+            => ExistingType?.BaseType is {} sourceBaseType && sourceBaseType.SpecialType != SpecialType.System_ValueType && sourceBaseType.SpecialType != SpecialType.System_Object ? sourceBaseType : null;
+
         protected override IEnumerable<ObjectTypeProperty> BuildProperties()
-        {
-            foreach (var field in Fields)
-                yield return new ObjectTypeProperty(field, Fields.GetInputByField(field), this);
-        }
+            => Fields.Select(f => new ObjectTypeProperty(f, Fields.GetInputByField(f)));
 
         protected override IEnumerable<ObjectTypeConstructor> BuildConstructors()
         {
-            yield return InitializationConstructor;
-            if (SerializationConstructor != InitializationConstructor)
+            if (!_inputModel.IsUnknownDiscriminatorModel)
+            {
+                yield return InitializationConstructor;
+            }
+
+            if (SerializationConstructor != InitializationConstructor && (!Configuration.Generation1ConvenienceClient || IncludeDeserializer))
+            {
                 yield return SerializationConstructor;
+            }
         }
 
         protected override bool EnsureHasJsonSerialization()
         {
             if (IsPropertyBag)
                 return false;
-            return true;
+            return _inputModelSerialization.Json;
         }
 
-        protected override bool EnsureHasXmlSerialization()
-        {
-            return false;
-        }
+        protected override bool EnsureHasXmlSerialization() => _inputModelSerialization.Xml is not null;
```

Comment:
I am thinking that do we really need these two flags `HasJsonSerialization` and `HasXmlSerialization`?
The `JsonSerialization` being `null` should perfectly work to indicate that this model does not have a JsonSerialization.
From the code perspective, these two flags are not really used a lot - we check if they have json or xml serialization in the serialization writer by checking if `JsonSerialization` is null.
I am trying in my public serialization PR to remove these flags

Reply:
Unless there is a circular dependency somewhere.

Reply:
it turns out there is not any - I have removed them in my PR